### DISCUSSION TITLE
Interface-ify v.pref

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [vlang/v#11872] [Ideas]
I would love to try to build a specific version of the V compiler targeting non-standard devices like consoles running homebrew. However, since I would be the only one maintaining corresponding vlib and v.pref settings for that, it wouldn't make total sense to merge this into master. Instead, I have thought of writing a different V frontend entirely for that purpose, that simply reuses existing generators and the likes but implements other operating systems and platforms.

One of the most "attractive" places to start is obviously `v.pref` since it defines the OS strings that are used for compile time defines and for picking up files to be compiled (`x_windows.v` vs `v_ios.v` for instance). So I was hoping to request that the infrastructure would also allow to take a different implementation of this module as an argument so it can safely be extended or even re-written and modified for this purpose without interfering with the V master, and thus containing all resulting issues within a spearate project that V mainline doesn't have to consider.

You think this would be possible? Thank you very much!

**Comment by spytheman** [comment]
The structures and enums defined in v.pref, are used by many other compiler modules, and I doubt it would be easy to make it extendable at the language level, also adding indirection for something, that is used as frequently as the .pref structs can have performance implications.

**Comment by spytheman** [comment]
Would it not be easier if you instead contribute your code in the main compiler?

**Comment by spytheman** [comment]
> since I would be the only one maintaining corresponding vlib and v.pref settings for that

You never know ... There may be other console users, that will share the maintenance :-) .

**Comment by IngwiePhoenix** [comment]
It would be easy - but there are a lot of specialized toolchains and other information.

Here are two examples: The GBA and PlayStation Vita.

- The GBA toolchain is from DevkitPRO (devkitARM) and has a variadic installation location that is reflected by `$DEVKITARM` and `$DEVKITPRO`.
- The Vita has not one, but two SDKs with two different approaches (reverse engeneered open-source reimplementation vs. reverse engeneered and binary compatible). Both have their pros and cons (custom GL impl versus "official" EGL impl)
- The GameBoy has no GL implementation, at all.
- Both use completely different memory allocation although being both based on ARM. The GBA is "bare" and does not go through a kernel, whilst the Vita, does.
- These differences would result in many additional files placed in `vlib/builtin/`.
- There would be a plethora of additions in `v.pref` and `v.builder` to accomodate each of these targets; and I am only mentioning two. I actually am planning five.

Adding all that to mainline V would result in a lot of code that is quite often not needed by anyone but a few people at best. I still want to go through with that, but I would also love to not pollute the main repository with seldom used code that isn't maintained as often or as "good" as the mainline code for the major platforms is.

Hence my idea to separate all these things out, host it in a different repo and possibly make a frontend that can cater towards the difference in all these toolchains as it adds other paths to the standard lookup path and has additional switches to pick the toolchain to use, as well as other small information (for instance, is the program compiled for the Vita a usermode binary or kernelmode?).

That is why I would prefer to not contribute it upstream; not to hold anything back, but rather to not pollute the mainline repository with things that only a very minority of people would actually want or need.

**Comment by dumblob** [comment]
Wouldn't simply a "technical" separation using git submodules be sufficient for now (to not slow you down @IngwiePhoenix ) thus buying us the time to decide later whether it shall become a fixed part of the *V* repo or not (e.g. once we **actually** see how much code it is and how much of it is suitable for mainline instead of guessing)?
